File: DonRyu/2022-09-21-Search in Rotated Sorted Array.js
//내가 한것
const search = (nums, target) => {
  let MapTable = new Map()
  for (let a = 0; a < nums.length; a++) {
    MapTable.set(nums[a], a)
  }
  if (MapTable.has(target)) {
    return MapTable.get(target)
  } else {
    return -1
  }
}
console.log(search([4, 5, 6, 7, 0, 1, 2], 4))

const search = (nums, target) => {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    let mid = Math.floor(left + (right - left) / 2)
    if (nums[mid] === target) return mid;
    if (nums[left] <= nums[mid]) {
      if (nums[left] <= target && target < nums[mid]) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    } else {
      if (nums[mid] < target && target <= nums[right]) {
        left = mid + 1
      } else {
        right = mid - 1
      }
    }
  }

  return -1
}

File: BenKim/2022-09-21-search_in_rotated_sorted_array.py
def searchInRotatedSortedArray(nums, target):
    left = 0
    right = len(nums) - 1
    while left < right:
        middle = (left + right) // 2


        # 오름차순 내림차순
        # 타겟값보다 큰지 작은지

        # 중간기준 오름차순
        if nums[middle] < nums[right] and nums[right] < target:
            left = middle
        else:
            middle = right